### PR TITLE
Add property `host` to `JavaDebugOptions`: support debugging via network

### LINKED ISSUE #20644: JVM 9+ No Way to Enable Remote Debugging
### Expected Behavior
In JDK 9, the agentlib options for address changed from requiring simply a port number to allowing a full address (host and port) to be specified.

This change caused the following debug string:

```
-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005
```

to only bind to 127.0.0.1 in JDK 9, compared to JDK 8 which would bind to 0.0.0.0.

The correct format to keep this behavior is 

```
-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005
```

Sources (couldn't find an official source):

https://www.baeldung.com/java-application-remote-debugging#from-java9
https://www.jetbrains.com/help/idea/debug-a-java-application-using-a-dockerfile.html#create-remote-debug-config

While most people are debugging locally, this becomes a problem especially when using Docker - which needs 0.0.0.0 to be bound to allow the host to communicate over the internal docker network.

### Current Behavior
Using the `--jvm-debug` parameter sets 

`-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005`

and only the server, suspend, and port can be overriden, instead of the whole address. In addition, port is an Integer, so you can't specify `*:5005`

### Context

Workaround:

```groovy
tasks {
    run {
        jvmArgs = ["-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005"]
    }
}
```

Suggested solution:

Add a new property to https://docs.gradle.org/current/javadoc/org/gradle/process/JavaDebugOptions.html to specify the address.
https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/core/src/main/java/org/gradle/process/internal/JvmOptions.java#L153

```java
        if (debugOptions.getEnabled().get()) {
            boolean server = debugOptions.getServer().get();
            boolean suspend = debugOptions.getSuspend().get();
            int port = debugOptions.getPort().get();
            String address = debugOptions.getAddress().get()
            args.add("-agentlib:jdwp=transport=dt_socket,server=" + (server ? 'y' : 'n') + ",suspend=" + (suspend ? 'y' : 'n') + ",address=" + (address != null ? "address" + ":" : "") + port);
        }
```

Alternatively, the address could take priority over the port.

```java
        if (debugOptions.getEnabled().get()) {
            boolean server = debugOptions.getServer().get();
            boolean suspend = debugOptions.getSuspend().get();
            int port = debugOptions.getPort().get();
            String address = debugOptions.getAddress().get()
            args.add("-agentlib:jdwp=transport=dt_socket,server=" + (server ? 'y' : 'n') + ",suspend=" + (suspend ? 'y' : 'n') + ",address=" + (address != null ? "address" + ":" : port));
        }
```

### PR DESCRIPTION
**Pull request by h0tk3y**
In Java 9 and above, if the JDWP arguments only include the port and not the host name, the debugger server will only listen on the loopback network interface, and it won't accept connections via network.

Debugging remotely with explicitly specified host address was only possible by passing the JDWP arguments as a string, which is a UX issue.
Adding a property for specifying the host address to `JavaDebugOptions` fixes that.

Also add daemon parameter property `org.gradle.debug.host`. 
This simplifies running the daemon in a way that it is debuggable remotely via network, as opposed to the loopback interface.

Fixes #20644

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/process/JavaDebugOptions.java
Comment (ljacomet) [inline subprojects/core-api/src/main/java/org/gradle/process/JavaDebugOptions.java:37]: That sentence is weird.
What about:
> The host on which to listen for the debugger.

I would also add that `*` can be used to listen on all interfaces.

Comment (ljacomet) [reply subprojects/core-api/src/main/java/org/gradle/process/JavaDebugOptions.java:37]: I would actually copy the content you have in the documentation for the properties.

**Review by ljacomet — APPROVED** [review]
Looks good, only a minor comment on the Javadoc, see below.

Can be merged once the javadoc is approved.

File: subprojects/core-api/src/main/java/org/gradle/process/JavaDebugOptions.java
Comment (h0tk3y) [reply subprojects/core-api/src/main/java/org/gradle/process/JavaDebugOptions.java]: @ljacomet Done, please take a look.

Comment (h0tk3y) [reply subprojects/core-api/src/main/java/org/gradle/process/JavaDebugOptions.java]: Since this host name is used both in the server mode and in the client mode, it can mean either the host to listen on or the host to connect to, so mentioning just the host on which to listen doesn't feel enough. I followed your suggestion and copied the property docs here. Assuming that you approved that one, I'm merging this now.

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Pull request merged**
